core/watcher/eventqueue/uuid.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package eventqueue
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/changestream"
+)
+
+// UUIDWatcher watches for changes to a database table.
+// Any time rows change in the watched table, the changed UUIDs are emitted.
+type UUIDWatcher struct {
+	*BaseWatcher
+
+	out       chan []string
+	tableName string
+	selectAll string
+}
+
+// NewUUIDWatcher returns a new watcher that receives changes from the
+// input base watcher's db/queue when rows in the input table change.
+func NewUUIDWatcher(base *BaseWatcher, tableName string) *UUIDWatcher {
+	w := &UUIDWatcher{
+		BaseWatcher: base,
+		out:         make(chan []string),
+		tableName:   tableName,
+		selectAll:   fmt.Sprintf("SELECT uuid FROM %s", tableName),
+	}
+
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Changes returns the channel on which the UUIDs for
+// changed rows are sent to downstream consumers.
+func (w *UUIDWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+func (w *UUIDWatcher) loop() error {
+	subscription, err := w.eventQueue.Subscribe(
+		changestream.Namespace(w.tableName, changestream.Create|changestream.Update|changestream.Delete))
+	if err != nil {
+		return errors.Annotatef(err, "subscribing to namespace %q", w.tableName)
+	}
+	defer subscription.Unsubscribe()
+
+	changes, err := w.getInitialState()
+	if err != nil {
+		return errors.Annotate(err, "retrieving initial watcher state")
+	}
+
+	// By reassigning the in and out channels, we effectively ticktock between
+	// read mode and dispatch mode. This ensures we always dispatch deltas that
+	// we received before reading more, and every channel read/write is guarded
+	// by checks of the tomb and subscription liveness.
+	// We begin in dispatch mode in order to send the initial state.
+	var in <-chan []changestream.ChangeEvent
+	out := w.out
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-subscription.Done():
+			return ErrSubscriptionClosed
+		case subChanges, ok := <-in:
+			if !ok {
+				w.logger.Debugf("change channel closed for %q; terminating watcher", w.tableName)
+				return nil
+			}
+
+			// We have changes. Tick over to dispatch mode.
+			changes = transform.Slice(subChanges, func(c changestream.ChangeEvent) string { return c.ChangedUUID() })
+			in = nil
+			out = w.out
+		case out <- changes:
+			// We have dispatched. Tick over to read mode.
+			in = subscription.Changes()
```

Comment:
I know we discussed this previously, but want to be aware of people who may come back to read this. If you do not read this quickly enough, you will be unsubscribed and the watcher should be restarted. It is then up to the watcher to start with an initial state and start afresh. A singular watcher should not ground the system to a halt. In the current implementation, all messages over the [deadline are dropped](https://github.com/juju/juju/blob/7a0c6fbc9ce4881014d298896c827c4e31857f0d/state/watcher/hubwatcher.go#L473-L476). No initial state is re-read, which means watchers are potentially dealing with bad data. The design of this new watcher and subscription model aims to solve that deficiency.

Reply:
@SimonRichardson you mentioned this, but I don't see anything in this code that handles it. (unless that is subscription.Done()), where are timeouts actually tracked?

It seems like *something* would need to let the subscription know that it read more content (which we had been blocked on).
What I see here, is that we do call .Changes() which could be "and we sent them some more data", but it is unclear if there are no messages to send, whether the subscriptions would expire due to inactivity. If we didn't want that, then we would need a "let you know that we have some more data to send... and you have received it".

Reply:
> unless that is subscription.Done()

it's 100% this.

Reply:
Maybe `Done` should be renamed to something more explicit. I'm happy to rename it. It needs to handle the case where the subscription was unsubscribed either because:

 1. It was unsubscribed because the user of the subscription was done (gracefully).
 2. It was timed out and forcibly removed.
 3. The worker was told to shut down, so the catacomb and all subscription tombs were killed.

Currently, we don't tell the subscription what the cause of the termination was, but maybe we could do?